File: src/Application/Services/Accountant/SyncService.cs
using System.Linq;
using System.Threading.Tasks;
using Application.Contracts.Accountant.Sync;
using Application.Contracts.Accountant.Sync.Models;
using AutoMapper;
using Domain.Entities.Accountant;

namespace Application.Services.Accountant;

public class SyncService : ISyncService
{
    private readonly ISyncRepository _syncRepository;
    private readonly IMapper _mapper;

    public SyncService(
        ISyncRepository syncRepository,
        IMapper mapper)
    {
        _syncRepository = syncRepository;
        _mapper = mapper;
    }

    public async Task<SyncedEntityIds> SyncEntitiesAsync(SyncEntities model)
    {
        var accounts = model.Accounts.Select(x => _mapper.Map<Account>(x)).ToList();
        var categories = model.Categories.Select(x => _mapper.Map<Category>(x)).ToList();
        var transactions = model.Transactions.Select(x => _mapper.Map<Transaction>(x)).ToList();
        var upcomingExpenses = model.UpcomingExpenses.Select(x => _mapper.Map<UpcomingExpense>(x)).ToList();
        var debts = model.Debts.Select(x => _mapper.Map<Debt>(x)).ToList();

        await _syncRepository.SyncAsync(accounts, categories, transactions, upcomingExpenses, debts);

        var accountIds = accounts.Select(x => x.Id).ToArray();
        var categoryIds = categories.Select(x => x.Id).ToArray();
        var transactionIds = transactions.Select(x => x.Id).ToArray();
        var upcomingExpenseIds = upcomingExpenses.Select(x => x.Id).ToArray();
        var debtIds = debts.Select(x => x.Id).ToArray();

        return new SyncedEntityIds(accountIds, categoryIds, transactionIds, upcomingExpenseIds, debtIds);
    }
}

File: src/Accountant.Application.Fs/Services/SyncService.fs
namespace Accountant.Application.Fs.Services

open Domain.Accountant
open Accountant.Application.Fs.Models.Sync

module SyncService =
    let mapEntitiesForSync (userId: int) (model: SyncEntities) =
        let accounts =
            model.Accounts
            |> Seq.map (fun x ->
                Account(
                    Id = x.Id,
                    UserId = userId,
                    Name = x.Name,
                    IsMain = x.IsMain,
                    Currency = x.Currency,
                    StockPrice = x.StockPrice,
                    CreatedDate = x.CreatedDate,
                    ModifiedDate = x.ModifiedDate
                ))

        let categories =
            model.Categories
            |> Seq.map (fun x ->
                Category(
                    Id = x.Id,
                    UserId = userId,
                    ParentId = x.ParentId,
                    Name = x.Name,
                    Type = x.Type,
                    GenerateUpcomingExpense = x.GenerateUpcomingExpense,
                    IsTax = x.IsTax,
                    CreatedDate = x.CreatedDate,
                    ModifiedDate = x.ModifiedDate
                ))

        let transactions =
            model.Transactions
            |> Seq.map (fun x ->
                Transaction(
                    Id = x.Id,
                    FromAccountId = x.FromAccountId,
                    ToAccountId = x.ToAccountId,
                    CategoryId = x.CategoryId,
                    Amount = x.Amount,
                    FromStocks = x.FromStocks,
                    ToStocks = x.ToStocks,
                    Currency = x.Currency,
                    Description = x.Description,
                    Date = x.Date,
                    IsEncrypted = x.IsEncrypted,
                    EncryptedDescription = x.EncryptedDescription,
                    Salt = x.Salt,
                    Nonce = x.Nonce,
                    Generated = x.Generated,
                    CreatedDate = x.CreatedDate,
                    ModifiedDate = x.ModifiedDate
                ))

        let upcomingExpenses =
            model.UpcomingExpenses
            |> Seq.map (fun x ->
                UpcomingExpense(
                    Id = x.Id,
                    UserId = userId,
                    CategoryId = x.CategoryId,
                    Amount = x.Amount,
                    Currency = x.Currency,
                    Description = x.Description,
                    Date = x.Date,
                    Generated = x.Generated,
                    CreatedDate = x.CreatedDate,
                    ModifiedDate = x.ModifiedDate
                ))

        let debts =
            model.Debts
            |> Seq.map (fun x ->
                Debt(
                    Id = x.Id,
                    UserId = userId,
                    Person = x.Person.Trim(),
                    Amount = x.Amount,
                    Currency = x.Currency,
                    UserIsDebtor = x.UserIsDebtor,
                    Description = x.Description,
                    CreatedDate = x.CreatedDate,
                    ModifiedDate = x.ModifiedDate
                ))

        let automaticTransactions =
            model.AutomaticTransactions
            |> Seq.map (fun x ->
                AutomaticTransaction(
                    Id = x.Id,
                    UserId = userId,
                    IsDeposit = x.IsDeposit,
                    CategoryId = x.CategoryId,
                    Amount = x.Amount,
                    Currency = x.Currency,
                    Description = x.Description,
                    DayInMonth = x.DayInMonth,
                    CreatedDate = x.CreatedDate,
                    ModifiedDate = x.ModifiedDate
                ))

        (accounts, categories, transactions, upcomingExpenses, debts, automaticTransactions)
